File: 1636/solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] frequencySort(int[] nums) {
        Map<Integer, Integer> counts = new HashMap<>();
        Integer[] ints = new Integer[nums.length];
        for (int i = 0; i < nums.length; i++) {
            ints[i] = nums[i];
            counts.put(nums[i], counts.getOrDefault(nums[i], 0) + 1);
        }

        Arrays.sort(ints, (a, b) -> {
            int freqA = counts.get(a);
            int freqB = counts.get(b);
            if (freqA == freqB) {
                return b - a;
            } else {
                return freqA - freqB;
            }
        });

        int[] res = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            res[i] = ints[i];
        }
        return res;
    }
}

// time O(nlogn)
// space O(n)

File: 1636/solution.ts
function frequencySort(nums: number[]): number[] {
  const counts = new Map<number, number>();
  for (const num of nums) {
    counts.set(num, (counts.get(num) || 0) + 1);
  }
  nums.sort((a, b) => {
    return counts.get(a) !== counts.get(b)
      ? counts.get(a)! - counts.get(b)!
      : b - a;
  });
  return nums;
}
